### PR TITLE
Add cascade to `dag_tag` to `dag` foreignkey

### LINKED ISSUE #23206: UI shows Foreign Key Error when deleting a dag
### Apache Airflow version

2.3.0b1 (pre-release)

### What happened

I tried to delete a dag from the grid view, and I saw this instead

```
Ooops!
Something bad has happened.
...
Python version: 3.7.13
Airflow version: 2.3.0b1
Node: airflow-webserver-7c4f49f5dd-h74w2
-------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1706, in _execute_context
    cursor, statement, parameters, context
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 716, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.ForeignKeyViolation: update or delete on table "dag" violates foreign key constraint "dag_tag_dag_id_fkey" on table "dag_tag"
DETAIL:  Key (dag_id)=(core_todo) is still referenced from table "dag_tag".


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/www/auth.py", line 40, in decorated
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/www/decorators.py", line 80, in wrapper
    return f(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/www/views.py", line 1812, in delete
    delete_dag.delete_dag(dag_id)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/session.py", line 71, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/api/common/delete_dag.py", line 80, in delete_dag
    .delete(synchronize_session='fetch')
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3111, in delete
    execution_options={"synchronize_session": synchronize_session},
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 1670, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1520, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 314, in _execute_on_connection
    self, multiparams, params, execution_options
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1399, in _execute_clauseelement
    cache_hit=cache_hit,
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1749, in _execute_context
    e, statement, parameters, cursor, context
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1930, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1706, in _execute_context
    cursor, statement, parameters, context
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 716, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) update or delete on table "dag" violates foreign key constraint "dag_tag_dag_id_fkey" on table "dag_tag"
DETAIL:  Key (dag_id)=(core_todo) is still referenced from table "dag_tag".

[SQL: DELETE FROM dag WHERE dag.dag_id IN (%(dag_id_1_1)s) RETURNING dag.dag_id]
[parameters: {'dag_id_1_1': 'core_todo'}]
(Background on this error at: http://sqlalche.me/e/14/gkpj)
```

Also, here are the database pod logs:
```
│ 2022-04-25 01:42:14.185 GMT [155] STATEMENT:  INSERT INTO log (dttm, dag_id, task_id, map_index, event, execution_date, owner, extra) VALUES ('2022-04-25T01:42:14.178085+00:00'::timestamptz, NULL, NULL, NULL, 'cli_upgradedb', NULL, ' │
│ 2022-04-25 01:42:14.371 GMT [155] ERROR:  relation "connection" does not exist at character 55                                                                                                                                            │
│ 2022-04-25 01:42:14.371 GMT [155] STATEMENT:  SELECT connection.conn_id AS connection_conn_id                                                                                                                                             │
│     FROM connection GROUP BY connection.conn_id                                                                                                                                                                                           │
│     HAVING count(*) > 1                                                                                                                                                                                                                   │
│ 2022-04-25 01:42:14.372 GMT [155] ERROR:  relation "connection" does not exist at character 55                                                                                                                                            │
│ 2022-04-25 01:42:14.372 GMT [155] STATEMENT:  SELECT connection.conn_id AS connection_conn_id                                                                                                                                             │
│     FROM connection                                                                                                                                                                                                                       │
│     WHERE connection.conn_type IS NULL                                                                                                                                                                                                    │
│ 2022-04-25 01:42:16.489 GMT [158] ERROR:  relation "log" does not exist at character 13                                                                                                                                                   │
│ 2022-04-25 01:42:16.489 GMT [158] STATEMENT:  INSERT INTO log (dttm, dag_id, task_id, map_index, event, execution_date, owner, extra) VALUES ('2022-04-25T01:42:16.482543+00:00'::timestamptz, NULL, NULL, NULL, 'cli_check', NULL, 'airf │
│ 2022-04-25 01:42:17.917 GMT [160] ERROR:  column "map_index" of relation "log" does not exist at character 41                                                                                                                             │
│ 2022-04-25 01:42:17.917 GMT [160] STATEMENT:  INSERT INTO log (dttm, dag_id, task_id, map_index, event, execution_date, owner, extra) VALUES ('2022-04-25T01:42:17.910396+00:00'::timestamptz, NULL, NULL, NULL, 'cli_flower', NULL, 'air │
│ 2022-04-25 03:18:33.631 GMT [24494] ERROR:  update or delete on table "dag" violates foreign key constraint "dag_tag_dag_id_fkey" on table "dag_tag"                                                                                      │
│ 2022-04-25 03:18:33.631 GMT [24494] DETAIL:  Key (dag_id)=(core_todo) is still referenced from table "dag_tag".                                                                                                                           │
│ 2022-04-25 03:18:33.631 GMT [24494] STATEMENT:  DELETE FROM dag WHERE dag.dag_id IN ('core_todo') RETURNING dag.dag_id                                                                                                                    │
│ 2022-04-25 03:31:18.858 GMT [24760] ERROR:  update or delete on table "dag" violates foreign key constraint "dag_tag_dag_id_fkey" on table "dag_tag"                                                                                      │
│ 2022-04-25 03:31:18.858 GMT [24760] DETAIL:  Key (dag_id)=(core_todo) is still referenced from table "dag_tag".                                                                                                                           │
│ 2022-04-25 03:31:18.858 GMT [24760] STATEMENT:  DELETE FROM dag WHERE dag.dag_id IN ('core_todo') RETURNING dag.dag_id    
```

### What you think should happen instead

The dag gets deleted, no error

### How to reproduce

I'm not sure if I can replicate it, but I'll report back here if I can.  So far as I remember the steps were:

1. run a (large) dag
2. the dag failed for unrelated reasons
3. delete the dag from the grid view
4. see error page


### Operating System

kubernetes/debian

### Versions of Apache Airflow Providers

n/a

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

Deployed via helm into a microk8s cluster, which was running in a VM, which was deployed by CircleCI.  

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #23414: airflow db clean - Dag cleanup won't run if dag is tagged
### Apache Airflow version

2.3.0 (latest released)

### What happened

When running `airflow db clean`, if a to-be-cleaned dag is also tagged, a foreign key constraint in dag_tag is violated. Full error:

``` 
sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) update or delete on table "dag" violates foreign key constraint "dag_tag_dag_id_fkey" on table "dag_tag"
DETAIL:  Key (dag_id)=(some-dag-id-here) is still referenced from table "dag_tag".
```

### What you think should happen instead

_No response_

### How to reproduce

_No response_

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==3.3.0
apache-airflow-providers-cncf-kubernetes==4.0.1
apache-airflow-providers-ftp==2.1.2
apache-airflow-providers-http==2.1.2
apache-airflow-providers-imap==2.2.3
apache-airflow-providers-microsoft-mssql==2.1.3
apache-airflow-providers-oracle==2.2.3
apache-airflow-providers-postgres==4.1.0
apache-airflow-providers-samba==3.0.4
apache-airflow-providers-slack==4.2.3
apache-airflow-providers-sqlite==2.1.3
apache-airflow-providers-ssh==2.4.3

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ephraimbuddy**
Bulk delete does not work if the cascade behaviour of a foreignkey
is set on python side(relationship configuration). To allow bulk delete of dags
we need to setup cascade deletion in the DB.

The warning on query.delete at
https://docs.sqlalchemy.org/en/14/orm/session_basics.html#selecting-a-synchronization-strategy
stated that:

The operations do not offer in-Python cascading of relationships - it is assumed that ON UPDATE CASCADE and/or ON DELETE CASCADE is configured for any foreign key references which require it, otherwise the database may emit an integrity violation if foreign key references are being enforced.

Another alternative is avoiding bulk delete of dags but I prefer we support bulk deletes.

This will break offline sql generation for mssql(already broken before now :) ). Also, since there's only one foreign key
in `dag_tag` table, I assume that the foreign key would be named `dag_tag_ibfk_1` in `mysql`. This
avoided having to query the db for the name.

The foreignkey is explicitly named now, would be easy for future upgrades

Closes: #23414, Closes: #23206

### CODE REVIEW

**Comment by ephraimbuddy** [comment]
You can test this by running a dag with tags and then deleting it with the `airflow db clean` command

**Comment by uranusjr** [comment]
Looks like SQLite needs some additional work

File: airflow/migrations/utils.py [hunk 1]
```diff
@@ -35,7 +35,8 @@ def get_mssql_table_constraints(conn, table_name):
      FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS tc
      JOIN INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE AS ccu ON ccu.CONSTRAINT_NAME = tc.CONSTRAINT_NAME
      WHERE tc.TABLE_NAME = '{table_name}' AND
-     (tc.CONSTRAINT_TYPE = 'PRIMARY KEY' or UPPER(tc.CONSTRAINT_TYPE) = 'UNIQUE')
+     (tc.CONSTRAINT_TYPE = 'PRIMARY KEY' or UPPER(tc.CONSTRAINT_TYPE) = 'UNIQUE'
+     or UPPER(tc.CONSTRAINT_TYPE) = 'FOREIGN KEY')
```
Comment (uranusjr) [inline airflow/migrations/utils.py:39]: One of these is not upper-cased! But do you really need `UPPER`? From what I remember, other migrations don’t do this amd still work OK.

Comment (ephraimbuddy) [reply airflow/migrations/utils.py:39]: I don't think there's any other migration that tried dropping an unnamed foreign key in `MSSQL`. Most other foreign keys are named and easily dropped. This particular case is one of those unnamed foreign keys. There's no other way to drop unnamed foreign key in MSSQL other than using this query or using sqlalchemy `inspect`.

For the one that's not uppercased, it has been like that and working. My addition to the code is that of foreign key.

Comment (uranusjr) [reply airflow/migrations/utils.py:39]: Hmm, and the other two can be in lower case? I would imagine we should be able to get rid of those `UPPER`;

Comment (ephraimbuddy) [reply airflow/migrations/utils.py:39]: Yea. Possible but it'll require editing released migrations. Where ever it was used, it compares to an upper case value. 
In my opinion we should keep it with the `UPPER` case

File: airflow/migrations/utils.py [hunk 2]
```diff
@@ -35,7 +35,8 @@ def get_mssql_table_constraints(conn, table_name):
      FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS tc
      JOIN INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE AS ccu ON ccu.CONSTRAINT_NAME = tc.CONSTRAINT_NAME
      WHERE tc.TABLE_NAME = '{table_name}' AND
```
Comment (ashb) [inline airflow/migrations/utils.py:37]: 😱  This should be a placeholder/param pass to execute, not built in to the string. (Separate to this PR)

**Review by jedcunningham — APPROVED** [review]

**Review by norm — APPROVED** [review]

**Review by norm — APPROVED** [review]

**Review by ashb — APPROVED** [review]

**Pull request merged**